PS20b Support + Parallax Occlusion + PBR Common header for function sharing

sears2424 opened the pull request:
1. PS2.0b Support (me , bonjorno7)
2. Parallax Occlusion Mapping (me , thexa4)
3. Test Mod template (Momentum Mod team, bonjorno7, Cabbage McGravel)
4. Moved functions into a header pbr_common_ps2_3_x.h for consistency. between 2.0b and 3.0 (me)

sears2424 committed:
PBR PS 2.0 Support

Props to Bonjorno for providing the test assets

sears2424 committed:
Created a PBR common header for 3.0 and 2.0b shaders to share functions with

sears2424 committed:
added Parallax Occlusion Mapping to brushes.

ps 3.0 only

sears2424 committed:
used $parallax instead of $useparallax, improved the demo files.

sears2424 committed:
Delete game_shader_dx9.pdb

thexa4 commented on mp/src/materialsystem/stdshaders/game_shader_dx9_hl2mp.vpc:
```diff
@@ -16,13 +16,15 @@ $Project "Shaders (HL2MP)"
 		{
 			$File	"pbr_ps30.fxc"
 			$File	"pbr_vs30.fxc"
+			$File	"pbr_ps20b.fxc"
```
Shouldn't the .h files be here as well?

thexa4 commented on mp/src/materialsystem/stdshaders/pbr_common_ps30_fxc.h:
```diff
@@ -0,0 +1,222 @@
+//===================== Copyright (c) Valve Corporation. All Rights Reserved. ======================
```
Should probably remove this copyright statement

thexa4 commented on mp/src/materialsystem/stdshaders/pbr_dx9.cpp:
```diff
@@ -30,12 +33,16 @@ const Sampler_t SAMPLER_SPECULAR = SHADER_SAMPLER12;
 // Convars
 static ConVar mat_fullbright("mat_fullbright", "0", FCVAR_CHEAT);
 static ConVar mat_specular("mat_specular", "1", FCVAR_CHEAT);
+static ConVar mat_pbr_force_20b("mat_pbr_force_20b", "0");
```
This should be `FCVAR_CHEAT`

thexa4 commented on mp/src/materialsystem/stdshaders/pbr_dx9.cpp:
```diff
@@ -30,12 +33,16 @@ const Sampler_t SAMPLER_SPECULAR = SHADER_SAMPLER12;
+static ConVar mat_pbr_force_20b("mat_pbr_force_20b", "0");
+
+#define PBRPARALLAX
+static ConVar mat_parallax_height("mat_parallax_height", "0");
```
This should be `FCVAR_CHEAT`. It doesn't seem to be used at the moment, can it be removed?

sears2424 replied:
yeah, forgot to remove that, just a debug CVar

thexa4 commented on mp/src/materialsystem/stdshaders/pbr_dx9.cpp:
```diff
@@ -68,6 +79,10 @@ BEGIN_VS_SHADER(PBR, "PBR shader")
 		SHADER_PARAM(SPECULARTEXTURE, SHADER_PARAM_TYPE_TEXTURE, "", "Specular F0 RGB map");
+#ifdef PBRPARALLAX
+		SHADER_PARAM(PARALLAX, SHADER_PARAM_TYPE_BOOL, "0", "Use Parallax Occlusion Mapping.");
+		SHADER_PARAM(PARALLAXHEIGHT, SHADER_PARAM_TYPE_FLOAT, "0.5", "Maximum Height of the Parallax Map");
```
Is 0.5 a good default?

sears2424 replied:
maybe too high, depends on the heightmap. i want 0.5 since when you toggle $parallax on, you can see results immediately even without modifying the $parallaxheight ...

thexa4 commented on mp/src/materialsystem/stdshaders/pbr_dx9.cpp:
```diff
@@ -68,6 +79,10 @@ BEGIN_VS_SHADER(PBR, "PBR shader")
 		SHADER_PARAM(BUMPMAP, SHADER_PARAM_TYPE_TEXTURE, "", "Normal texture");
         SHADER_PARAM(USEENVAMBIENT, SHADER_PARAM_TYPE_BOOL, "0", "Use the cubemaps to compute ambient light.");
 		SHADER_PARAM(SPECULARTEXTURE, SHADER_PARAM_TYPE_TEXTURE, "", "Specular F0 RGB map");
+#ifdef PBRPARALLAX
```
Why does the parallax need to be IfDeffed?

sears2424 replied:
Habit... ill remove it if you desire not to have them

thexa4 commented on mp/src/materialsystem/stdshaders/pbr_dx9.cpp:
```diff
@@ -271,24 +290,52 @@ BEGIN_VS_SHADER(PBR, "PBR shader")
 			else
 			{
 				// We need the position, surface normal, and vertex compression format
```
Comment no longer corresponds to the code below

thexa4 commented on mp/src/materialsystem/stdshaders/pbr_dx9.cpp:
```diff
@@ -271,24 +290,52 @@ BEGIN_VS_SHADER(PBR, "PBR shader")
+#ifdef PBRPARALLAX
+			const int useParallax = params[info.useParallax]->GetIntValue();
+#endif
+			if (!g_pHardwareConfig->SupportsShaderModel_3_0() || mat_pbr_force_20b.GetBool())
```
Why didn't you use the fallback system?

sears2424 replied:
i read VertexLitGeneric shader and this is how valve does it.

thexa4 commented on mp/src/materialsystem/stdshaders/pbr_dx9.cpp:
```diff
@@ -539,6 +607,15 @@ BEGIN_VS_SHADER(PBR, "PBR shader")
 
+#ifdef		PBRPARALLAX
+			float flParams[4] = { 0.0f, 0.0f, 0.0f, 0.0f };
+			flParams[0] = 50;
```
What are these constants?

sears2424 replied:
initialization.

thexa4 replied:
Can you document what the values mean? You're currently initializing 50, 50 which are unitless.

thexa4 replied:
This is no longer being used right?

thexa4 commented on mp/src/materialsystem/stdshaders/pbr_ps20b.fxc:
```diff
@@ -0,0 +1,219 @@
+//===================== Copyright (c) Valve Corporation. All Rights Reserved. ======================
```
Should probably remove the copyright header

thexa4 commented on mp/src/materialsystem/stdshaders/pbr_ps20b.fxc:
```diff
@@ -0,0 +1,219 @@
+	//float3 surfTangent;
+	//float3 surfBase;
+	//float flipSign;
+	//float3x3 normalBasis = compute_tangent_frame(surfNormal, i.worldPos, i.baseTexCoord, surfTangent, surfBase, flipSign);
```
Could you remove commented out code?

sears2424 replied:
alright

thexa4 commented on mp/src/materialsystem/stdshaders/pbr_ps20b.fxc:
```diff
@@ -0,0 +1,219 @@
+	float3 outgoingLightDirection = normalize(g_EyePos.xyz - i.worldPos); // Lo
+	float lightDirectionAngle = max(0, dot(normal, outgoingLightDirection)); // cosLo
+
+	float3 specularReflectionVector = 2.0 * lightDirectionAngle * normal - outgoingLightDirection; // Lr
```
Shouldn't most of this be in the common .h file as well?

sears2424 replied:
we can, right now i did the functions only since those will break the shaders the least.

thexa4 commented on mp/src/materialsystem/stdshaders/pbr_dx9.cpp:
```diff
@@ -271,24 +290,52 @@ BEGIN_VS_SHADER(PBR, "PBR shader")
 			{
 				// We need the position, surface normal, and vertex compression format
-                unsigned int flags = VERTEX_POSITION | VERTEX_NORMAL;
+				unsigned int flags = VERTEX_POSITION | VERTEX_NORMAL | VERTEX_TANGENT_S | VERTEX_TANGENT_T;
```
This is only needed for the ps2.0 version right?

sears2424 replied:
this is needed for parallax occlusion.

thexa4 commented on mp/src/materialsystem/stdshaders/pbr_vs20b.fxc:
```diff
@@ -0,0 +1,116 @@
+//===================== Copyright (c) Valve Corporation. All Rights Reserved. ======================
```
Copyright header should probably be removed

thexa4 commented on mp/src/materialsystem/stdshaders/pbr_vs30.fxc:
```diff
@@ -9,6 +9,7 @@
 //	DYNAMIC: "SKINNING"					"0..1"
 //  DYNAMIC: "LIGHTING_PREVIEW"			"0..1"
 //  DYNAMIC: "NUM_LIGHTS"				"0..4"
+//	STATIC:  "PARALLAXOCCLUSION"		"0..1"
```
Do we need a separate toggle for self shadowing?

sears2424 replied:
i guess we need it, i havent put the self-shadowing code yet since performance penalty is still too high.

sears2424 replied:
the shader compiles for approx. 12 minutes now. (i5 4460, GTX970) so we might not want to add a `//static: PARALLAXSHADOW` should we implement parallax shadows because it will double the combo count for a few lines of code.

thexa4 replied:
Compilation time is less of an issue than the runtime of the shadows. You can add a SKIP to not enable shadows if parallax isn't enabled. That should only increase the combos by 50%.

thexa4 commented on mp/src/materialsystem/stdshaders/pbr_vs30.fxc:
```diff
@@ -29,6 +30,14 @@ struct VS_INPUT
+//for parallax mapping..
+	float4 positionWS  				: POSITION;
+	float3 vNormalWS   				: NORMAL;
+	float3 vBinormalWS 				: BINORMAL;
```
The pbr shader uses a different normal basis than source usually does. Do we need these?

sears2424 replied:
we need that to calculate the local position. thats how i did it by getting the data from the engine itself, I made it that way so that we can have parallax occlusion on 2.0b too. (2.0b removed the compute_tangent_frame function)

thexa4 replied:
Can you use `compute_tangent_frame` while on 3.0 instead of the source ones?

sears2424 replied:
`compute_tangent_frame` happens after the `parallaxCorrect` function, and takes the modified texCoords as input. im not sure if i can do it before.

sears2424 committed:
renamed the common header to a better name

sears2424 commented:
I renamed the common header to a better name
pbr_common_ps2_3_x.h
because it was also used by 2.0b shaders

gdianaty commented:
my name is Graham Dianaty and I approve this message

sears2424 committed:
Moved TBN computation for ps3.0 on the pixel shader.

Minor renames, fixes and removal of #ifdefs

sears2424 committed:
remove parallax combo remnants on ps20b shader

thexa4 commented on mp/src/materialsystem/stdshaders/pbr_vs30.fxc:
```diff
@@ -66,10 +58,8 @@ VS_OUTPUT main( const VS_INPUT v )
-	float3x3 mWorldToTangent = float3x3( v.vTangentWS, v.vBinormalWS, v.vNormalWS );
-	o.vEyePosTS  = mul( mWorldToTangent, vViewWS  );
+	float3 vEyePosWS = cEyePos - v.vPos;
+	o.vEyePosWS = vEyePosWS;
```
Isn't the world eye position given as a constant to the pixel shader?

thexa4 commented on mp/src/materialsystem/stdshaders/pbr_vs30.fxc:
```diff
@@ -66,10 +58,8 @@ VS_OUTPUT main( const VS_INPUT v )
-	//Out.vViewWS = vViewWS;
-	float3x3 mWorldToTangent = float3x3( v.vTangentWS, v.vBinormalWS, v.vNormalWS );
-	o.vEyePosTS  = mul( mWorldToTangent, vViewWS  );
+	float3 vEyePosWS = cEyePos - v.vPos;
```
cEyePos should already be in world space. Maybe this needs a better name?

sears2424 committed:
completely removed parallax occlusion code on the vertex shader, renamed variables

sears2424 committed:
set the numLayers variable to be constant 20, then removed all constants on the cpp side.

sears2424 committed:
fixed PBR Parallax Occlusion

sears2424 committed:
fixed static meshes not using envAmbient, fixed demo materials to have emissiontexture

sears2424 committed:
added viewmodel PBR test gun, added $parallaxcenter vmt param, changed $parallaxheight to $parallaxdepth

sears2424 committed:
fixed flashlight behaviour in water.

sears2424 committed:
cleaned up ps2.0b code.

sears2424 committed:
compiled stuff

sears2424 committed:
fix cubemap brightness on ps20b shader

sears2424 committed:
compiled files

sears2424 committed:
Updated the parallax code to compile properly with valve's shadercompile.

TODO: create a variable range that we can use with ConVars.

sears2424 committed:
compiled files.

bonjorno7 committed:
Replace tabs with spaces

Don't forget to set VS Code to indent using spaces from now on

bonjorno7 committed:
Improve consistency of comments

Replaced the nonsensical comment about supporting both OGL and DX normal maps (which I wrote when I had no idea what I was doing) with one that actually makes sense.
Made sure another comment starts with a space because that's how we do things around here mkay.
Tried to make the comments about parallax more understandable, though my description of parallax center isn't very good idk

sears2424 committed:
Merge pull request #1 from bonjorno7/test-support-ps20b

Minor code cleanup

bonjorno7 committed:
Add important flags

This fixes prop_static being too dark!

bonjorno7 committed:
Improve comments

No need for the Valve copyright, right?
No need to mention bugs here, that's what github is for.
And then just some minor cleanup nonsense.

sears2424 committed:
Merge pull request #2 from bonjorno7/test-support-ps20b

Fix prop_static being dark

bonjorno7 committed:
Fix important flag

Turns out tangent spaces are necessary for lighting, even if you don't use them.

bonjorno7 committed:
Don't automatically enable UseEnvAmbient

Now that ambient cubes work on static props, I think it's best to use them, so they look consistent with dynamic props.

bonjorno7 committed:
Fix brushes being too bright

Why were we scaling this with PI anyway?

bonjorno7 committed:
Update compiled shaders

To save future generations the 30 minute compile time

sears2424 committed:
Merge pull request #3 from bonjorno7/test-support-ps20b

Fix lighting some more

sears2424 committed:
Add missing binaries

sears2424 committed:
Update Binaries

bonjorno7 committed:
Cleanup for Gocnak

So it'll be easier to merge future versions of this shader back into Momentum Mod

sears2424 committed:
Merge pull request #4 from bonjorno7/test-support-ps20b

Cleanup for Gocnak.

sears2424 committed:
Update test mod files
